File: c++/0116-populating-next-right-pointers-in-each-node.cpp
class Solution {
public:
    Node* connect(Node* root) {
        if(root == NULL || (root -> right == NULL && root -> left == NULL))
            return root;
        queue <Node*>q;
        q.push(root);

        while(!q.empty()) {
            int size = q.size();
            Node* temp = NULL;

            for(int i = 0 ; i < size ; i++) {
                Node* front = q.front();

                if(temp != NULL)
                    temp -> next = front;
                temp = front;

                if(front -> left) {
                    q.push(front -> left);
                    q.push(front -> right);
                }
                q.pop();
            }
        }
        return root;
    }
};

File: kotlin/0116-populating-next-right-pointers-in-each-node.kt
class Solution {
    fun connect(root: Node?): Node? {
        var cur = root
        var next = cur?.left

        while (cur != null && next != null) {
            cur?.left?.next = cur?.right
            if (cur?.next != null)
                cur?.right?.next = cur?.next?.left

            cur = cur?.next
            if (cur == null) {
                cur = next
                next = cur?.left
            }
        }
        
        return root
    }
}
